### ISSUE TITLE
How exactly does "rollingUpdateMinInstancesInService" work

### ISSUE THREAD

**Issue by drywheat** [coreos/kube-aws#283]
This tool allows us to configure autoscaling groups for both controller and worker nodes inside **cluster.yaml**, indicating a `minSize`, `maxSize` and `rollingUpdateMinInstancesInService`. The first two keys are straightforward, but the latter is less obvious to me since it refers to *behavior during state transition*.

Here are a couple examples of how I'm thinking this might work:

#### Sample configuration in the scenarios below

```yaml
controller:
  autoScalingGroup:
    minSize: 2
    maxSize: 6
    rollingUpdateMinInstancesInService: 2
```

### Scenario A

Two new controller instances are launched ( there's a total of 4 running now ). Once the two new controllers come online, are deemed **In Service** (health check, ack, etc...) and start accepting traffic, the older two instances are deregistered and decommissioned.

### Scenario B

**In Service** is loosely defined to mean a node has been provisioned. Two new controller instances are launched and the old controller nodes are removed immediately thereafter. This may result in a brief API service interruption and/or cluster failure. This means I should have set `rollingUpdateMinInstancesInService` to 4 in order to ensure the older, functional nodes stay in service during the update.

So, basically, I want to know if I should set `rollingUpdateMinInstancesInService` equal to `minSize`, `2 * minSize`, `minSize - 1`, or anything else that will ensure there are no service interruptions during an update.

**Comment by c-knowles** [comment]
Hi @drywheat, thanks for the question.

`rollingUpdateMinInstancesInService` directly corresponds to `MinInstancesInService` inside the [AWS AutoScalingRollingUpdate Policy](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html#cfn-attributes-updatepolicy-rollingupdate)

There's more info in the AWS docs linked above but in summary:
* For stack creation, the rolling update doesn't do anything that I know of.
* For stack update, the `MinInstancesInService` specifies the minimum number of instances in service prior to older instances being terminated. So in kube-aws, it's used during cluster updates.

The actual `In Service` definition is a little less well defined but quite a lot of work has gone into improving this in kube-aws recently to make it wait correctly for the instances. I'd recommend you take a look at the `.Experimental.WaitSignal.Enabled` flag which enables the newer logic to [wait for resources](https://github.com/coreos/kube-aws/blob/master/config/templates/stack-template.json#L73).

@mumoshu I think we could with documentation improvements here, possibly even change `rollingUpdateMinInstancesInService` to `rollingUpdate.minInstancesInService`.

**Comment by mumoshu** [comment]
@c-knowles Thanks!

Just curious but what is the intention of the suggested change to `rollingUpdate.minInstancesInService`? (Sounds cleaner but not exactly sure why the change is needed

For the doc improvement, could I just update https://github.com/coreos/kube-aws/blob/master/Documentation/kubernetes-on-aws-render.md to include your nice explanation quoted below?

> `rollingUpdateMinInstancesInService` directly corresponds to `MinInstancesInService` inside the [AWS AutoScalingRollingUpdate Policy](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html#cfn-attributes-updatepolicy-rollingupdate)
>
>There's more info in the AWS docs linked above but in summary:
>* For stack creation, the rolling update doesn't do anything that I know of.
>* For stack update, the `MinInstancesInService` specifies the minimum number of instances in service prior to older instances being terminated. So in kube-aws, it's used during cluster updates.

**Comment by c-knowles** [comment]
@mumoshu, the intention was to clarify the meaning and perhaps allow other customisation a to the rolling update later.

Feel free to use what I wrote, but remove the `that I know of` ;)

**Comment by drywheat** [comment]
The intelligent way to configure this is actually pretty clear to me now. I think a *General Recommendations* section on how to configure the rolling updates section might be useful in order to address others concerns: 

* HA during update  ( e.g. what happens if an AZ goes down during a rolling update? )
* Throughput ( i.e. how many instances do I need servicing requests during a rolling update? )
* Outline a couple of scenarios ( e.g. simple non-HA setup, twin AZ setup )

I'll close this for now.

**Comment by mumoshu** [comment]
Thanks for the feedback @drywheat :bow:
I've submitted an another issue to reflect your suggestion #458
